### PR TITLE
add ge2e and tacotron2_aishell3 example

### PR DESCRIPTION
**Pull request by iclementine**
1. add ge2e and tacotron2_aishell3 example;
2. switch to visualdl as visualizer;
3. extend tacotron2 to support global condition and tone;
4. extend tacotron2 to support alternative stop conditioning.
5. Add README for ge2e, aishell_tacotron;
6. other chanegs.

### CODE REVIEW

**Comment by iclementine** [comment]
Tp replace [PR #106](https://github.com/PaddlePaddle/Parakeet/pull/106), all the updates are included here. And further updates are posted here.

File: examples/ge2e/README.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
+# Speaker Encoder
+
+This experiment trains a speaker encoder with speaker verification as it is task. It is done as a part of the experiment of transfer learning from speaker verification to multispeaker text-to-speech synthesis, which can be find at [tacotron2_aishell3](../tacotron2_shell3). The trained speaker encoder is used to ectract utterance embedding from utterances.
```
Comment (ZeyuChen) [inline examples/ge2e/README.md]: `verification as it is task` 不是太理解这句英文

Comment (ZeyuChen) [reply examples/ge2e/README.md]: `ectract` -> `extract`

File: examples/ge2e/README.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+## Model
+
+The model used is this experiment is the speaker encoder with text independent speaker verification task in [GENERALIZED END-TO-END LOSS FOR SPEAKER VERIFICATION](https://arxiv.org/pdf/1710.10467.pdf). GE2E-softmax loss is used.
```
Comment (zh794390558) [inline examples/ge2e/README.md]: `The model used in this experiment`,

File: examples/ge2e/README.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+4. magicdata
+
+   A Mandarin Chinese multispeaker dataset ，[URL](https://www.openslr.org/68/) .
+
+If you want to use other datasets, you can also download and preprocess it as long as it meets the needs described below.
+
+## Preprocess Datasets
+
+Multispeaker datasets are used as training data, though the transcriptions are not used. To enlarge the amount of data used for training, several multispeaker datasets are combined. The preporcessed dataset s are organized in a file structure described below. The mel spectrogram of each utterance is save in `.npy` format. The dataset is 2-straitified (speaker-utterance). Since multiple datasets are combined, to avoid conflict in speaker id, dataset name is prepended to the speake ids.
```
Comment (zh794390558) [inline examples/ge2e/README.md]: `The preporcessed datasets are organized`

Comment (zh794390558) [reply examples/ge2e/README.md]: 2-stratified

File: examples/ge2e/README.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+`--device` is the device type to run the training, 'cpu' and 'gpu' are supported.
+
+`--nprocs` is the number of replicas to run in multiprocessing based parallel training。Currently multiprocessing based parallel training is only enabled when using 'gpu' as the devicde. `CUDA_VISIBLE_DEVICES` can be used to specify visible devices with cuda。
+
+Other options。
+
+`--config` is a `.yaml` config file used to override the default config(which is coded in `config.py`)。
+
+`--opts` is command line options to further override config files。It should be the last comman line options passed with multiple key-value pairs separated by spaces。
```
Comment (zh794390558) [inline examples/ge2e/README.md]: using `files.`

File: examples/ge2e/README.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+`--checkpoint_path` specifies the checkpoiont to load before training, extension is not included。A parameter file ( `.pdparams`) and an optimizer state file ( `.pdopt`) with the same name is used. This option has a higher priority than auto-resuming from the `--output` directory.
```
Comment (zh794390558) [inline examples/ge2e/README.md]: `included。`

File: examples/ge2e/README.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+`--checkpoint_path` is the path of the checkpoint to use, extension not included.
+
+`--pattern` is the wildcard pattern to filter audio files for inference, defaults to `*.wav`.
+
+`--device` `--opts` has the same meaning as the training script.
+
+## References
+
+1. [GENERALIZED END-TO-END LOSS FOR SPEAKER VERIFICATION](https://arxiv.org/pdf/1710.10467.pdf)
+2. [Transfer Learning from Speaker Verification toMultispeaker Text-To-Speech Synthesis](https://arxiv.org/pdf/1806.04558.pdf)
```
Comment (zh794390558) [inline examples/ge2e/README.md]: `toMultispeaker`

File: examples/ge2e/inference.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+def embed_utterance(processor, model, fpath_or_wav):
+    # audio processor
+    wav = processor.preprocess_wav(fpath_or_wav)
+    mel_partials = processor.extract_mel_partials(wav)
+
+    model.eval()
+    # speaker encoder
+    with paddle.no_grad():
+        mel_partials = paddle.to_tensor(mel_partials)
+        with paddle.no_grad():
```
Comment (zh794390558) [inline examples/ge2e/inference.py:37]: why double no_grad?

File: examples/ge2e/dataset_processors.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+
+def process_magicdata(processor,
+                      datasets_root,
+                      output_dir,
+                      skip_existing=False):
+    dataset_name = "magicdata/train"
+    dataset_root = datasets_root / dataset_name
+
+    speaker_dirs = list((dataset_root).glob("*"))
```
Comment (zh794390558) [inline examples/ge2e/dataset_processors.py:181]: useless () of dataset_root

File: examples/ge2e/dataset_processors.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
+from functools import partial
```
Comment (zh794390558) [inline examples/ge2e/dataset_processors.py:15]: no copyright

File: examples/ge2e/dataset_processors.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    # datastes root: a reference path to compute speaker_name
+    # we prepand dataset name to speaker_id becase we are mixing serveal
+    # multispeaker datasets together
+    speaker_name = "_".join(speaker_dir.relative_to(datasets_root).parts)
+    speaker_output_dir = output_dir / speaker_name
+    speaker_output_dir.mkdir(parents=True, exist_ok=True)
+
+    # load exsiting file set
+    sources_fpath = speaker_output_dir / "_sources.txt"
+    if sources_fpath.exists():
```
Comment (zh794390558) [inline examples/ge2e/dataset_processors.py:56]: existing_names = {} before if

File: examples/ge2e/audio_processor.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+    pcm_wave = struct.pack("%dh" % len(wav),
+                           *(np.round(wav * INT16_MAX)).astype(np.int16))
+
+    # Perform voice activation detection
+    voice_flags = []
+    vad = webrtcvad.Vad(mode=3)
+    for window_start in range(0, len(wav), samples_per_window):
+        window_end = window_start + samples_per_window
+        voice_flags.append(
+            vad.is_speech(pcm_wave[window_start * 2:window_end * 2],
```
Comment (zh794390558) [inline examples/ge2e/audio_processor.py]: hard code 2

File: examples/ge2e/preprocess.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
+import argparse
+from pathlib import Path
+from config import get_cfg_defaults
+from audio_processor import SpeakerVerificationPreprocessor
+from dataset_processors import (process_librispeech, process_voxceleb1,
```
Comment (zh794390558) [inline examples/ge2e/preprocess.py:19]: multi line import

File: examples/ge2e/preprocess.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        "the train set of these datastes will be used. Possible names: librispeech_other, "
+        "voxceleb1, voxceleb2, aidatatang_200zh, magicdata.")
+    parser.add_argument(
+        "--skip_existing",
+        action="store_true",
+        help=
+        "Whether to skip ouput files with the same name. Useful if this script was interrupted."
+    )
+    parser.add_argument(
+        "--no_trim",
```
Comment (zh794390558) [inline examples/ge2e/preprocess.py:46]: trim recommend

File: examples/ge2e/train.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+from paddle import distributed as dist
+from paddle.optimizer import Adam
+from paddle import DataParallel
+from paddle.io import DataLoader
+from paddle.nn.clip import ClipGradByGlobalNorm
+
+from parakeet.models.lstm_speaker_encoder import LSTMSpeakerEncoder
+from parakeet.training import ExperimentBase, default_argument_parser
+
+from speaker_verification_dataset import (MultiSpeakerMelDataset,
```
Comment (zh794390558) [inline examples/ge2e/train.py]: one import per line.

File: examples/ge2e/train.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        start = time.time()
+        batch = self.read_batch()
+        data_loader_time = time.time() - start
+
+        self.optimizer.clear_grad()
+        self.model.train()
+        specs = batch
+        loss, eer = self.model(specs, self.config.training.speakers_per_batch)
+        loss.backward()
+        self.model_core.do_gradient_ops()
```
Comment (zh794390558) [inline examples/ge2e/train.py:72]: what's this?

File: examples/tacotron2_aishell3/README_cn.md [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+python process_wav.py --input=<input> --output=<output> --alignment=<alignment>
+```
+
+默认 input, output, alignment 分别是 `~/datasets/aishell3/train/wav`, `~/datasets/aishell3/train/normalized_wav`, `~/datasets/aishell3/train/alignment`.
+
+处理结束后，会将处理好的音频保存在 `<output>` 文件夹中。
+
+### 转录文本处理
+
+把文本转换称为 phone 和 tone 的形式，并存储起来。值得注意的是，这里我们的处理和用于 montreal force aligner 的不一样。我们把声调分了出来。这是一个处理方式，当然也可以只做声母和韵母的切分。
```
Comment (zh794390558) [inline examples/tacotron2_aishell3/README_cn.md]: `把文本转换称为`

File: examples/transformer_tts/synthesize.py [hunk 17]
```diff
@@ -13,22 +13,20 @@
...
-from parakeet.utils import scheduler
-from parakeet.training.cli import default_argument_parser
-from parakeet.utils.display import add_attention_plots
+from parakeet.utils import display
 
 from config import get_cfg_defaults
 
 
-@paddle.fluid.dygraph.no_grad
+@paddle.no_grad()
```
Comment (zh794390558) [inline examples/transformer_tts/synthesize.py]: uesless

**Review by zh794390558 — COMMENTED** [review]
1. syllable 和 phone 的模块有点儿多，后面可以同一成一个。
2. 特征部分也不少，期待audio可以同一下。
3. wavenet 为啥删除了？

File: examples/ge2e/README.md [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+
+3. Aidatatang-200zh
+
+   A Mandarin Chinese multispeaker dataset ，[URL](https://www.openslr.org/62/) .
+
+4. magicdata
+
+   A Mandarin Chinese multispeaker dataset ，[URL](https://www.openslr.org/68/) .
+
+If you want to use other datasets, you can also download and preprocess it as long as it meets the needs described below.
```
Comment (ZeyuChen) [inline examples/ge2e/README.md]: `meets the needs described below.` -> `meets the requirements`

File: examples/ge2e/README.md [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+
+## Training
+
+When preprocessing is done, run the command below to train the mdoel.
+
+```bash
+python train.py --data=<data_path> --output=<output> --device="gpu" --nprocs=1
+```
+
+`--data` is the path to the preprocessed dataset。
```
Comment (ZeyuChen) [inline examples/ge2e/README.md]: 所有 --xxxx 之前建议加 * 或者+ ，文档呈现格式呈现更紧凑一些

File: examples/ge2e/audio_processor.py [hunk 20]
```diff
@@ -0,0 +1,221 @@
+from pathlib import Path
```
Comment (ZeyuChen) [inline examples/ge2e/audio_processor.py:15]: Add copyright and license.

**Review by ZeyuChen — COMMENTED** [review]
Few comments

File: examples/ge2e/README.md (hunk 2 above)
Comment (iclementine) [reply examples/ge2e/README.md]: syllable 和 phone 确实可以合成一个。目前是暂时写在 examples 里面，因为感觉通用性还不够。后续在 frontend 里面会有更完善的写法。

ge2e 的 audio 处理和其他模型都不太一样，包括它使用的 sample rate 以及 STFT 配置，而且里面还有一个 VAD 移除空白的过程。所以暂时是放在 example 目录下。我把它整合成一个模块。

音频处理相关的代码分了几层，
1. 第一层(`audio_processor.py`)是提供本实验需要用到的 audio processing functions; 
2. 第二层 (dataset_processors.py) 是用于多个 dataset 的处理函数；
3. 第三层(preprocess.py) 是用于直接运行的脚本。
当时这么分是为了把功能层级分开。

wavenet 在 0.2 版本中已经暂时被移除，可能是训练出了问题，训练达不到此前 checkpoint 的效果。（彻底 debug 之后会重新加回来。）

Comment (iclementine) [reply examples/ge2e/README.md]: > `The model used in this experiment`,

这个改好啦~

File: examples/ge2e/README.md (hunk 3 above)
Comment (iclementine) [reply examples/ge2e/README.md]: 好的，已经修改。

File: examples/ge2e/README.md (hunk 4 above)
Comment (iclementine) [reply examples/ge2e/README.md]: 已经修改。

File: examples/ge2e/README.md (hunk 5 above)
Comment (iclementine) [reply examples/ge2e/README.md]: 已经修改。可能是输入法状态的原因，输了好多中文句号~

File: examples/ge2e/README.md (hunk 6 above)
Comment (iclementine) [reply examples/ge2e/README.md]: 已经修改

File: examples/ge2e/preprocess.py (hunk 13 above)
Comment (iclementine) [reply examples/ge2e/preprocess.py:46]: --no_trim 默认值是 False 的

File: examples/ge2e/train.py (hunk 15 above)
Comment (iclementine) [reply examples/ge2e/train.py:72]: 模型中的一个特别的操作。缩放梯度。而且这个操作是在 global grad norm scale 之前。所以还不单纯等于自定义学习率。所以就这么做了。

File: examples/ge2e/train.py (hunk 14 above)
Comment (iclementine) [reply examples/ge2e/train.py]: 已经修改。

File: examples/transformer_tts/synthesize.py (hunk 17 above)
Comment (iclementine) [reply examples/transformer_tts/synthesize.py]: 这个仍然是有必要的，因为 eval() 并不会自动开启 no_grad。

File: examples/ge2e/README.md (hunk 1 above)
Comment (iclementine) [reply examples/ge2e/README.md]: > verification as it is task

应该是 verification as its task 的，写错了。

Comment (iclementine) [reply examples/ge2e/README.md]: 已经修改。

File: examples/ge2e/README.md (hunk 18 above)
Comment (iclementine) [reply examples/ge2e/README.md]: 已经修改。

File: examples/ge2e/README.md (hunk 19 above)
Comment (iclementine) [reply examples/ge2e/README.md]: 已经修改。

File: examples/ge2e/audio_processor.py (hunk 20 above)
Comment (iclementine) [reply examples/ge2e/audio_processor.py:15]: 已经添加。

File: examples/tacotron2_aishell3/README_cn.md (hunk 16 above)
Comment (iclementine) [reply examples/tacotron2_aishell3/README_cn.md]: 已经修改

File: examples/ge2e/train.py (hunk 15 above)
Comment (zh794390558) [reply examples/ge2e/train.py:57]: 这里没有用parallel model 做，不会有问题吗？

Comment (zh794390558) [reply examples/ge2e/train.py:72]: 这个不用 parallel model 做，多卡时候会对吗？

File: examples/transformer_tts/synthesize.py (hunk 17 above)
Comment (zh794390558) [reply examples/transformer_tts/synthesize.py]: 下面的代码里有 no_grad

Comment (iclementine) [reply examples/transformer_tts/synthesize.py]: 已经修改

File: examples/ge2e/train.py (hunk 15 above)
Comment (iclementine) [reply examples/ge2e/train.py:72]: 不用多卡的时候 model_core 就是 model，它们是一样的。
